src/coredump/coredump.c

```diff
@@ -1086,40 +1086,65 @@ static int process_socket(int fd) {
+                                }
+
+                                /* This is zero length message but it either doesn't carry a single descriptor,
+                                 * or it has more than one. This is a protocol violation so let's bail out. */
+                                cmsg_close_all(&mh);
+                                r = log_error_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                                    "Received zero length message with zero or more than one file descriptor(s).");
+                                goto finish;
+                        }
+
+                        /* Second zero length message might carry two file descriptors, coredump fd and mount tree fd. */
                         found = cmsg_find(&mh, SOL_SOCKET, SCM_RIGHTS, CMSG_LEN(sizeof(int) * 2));
                         if (found) {
                                 int fds[2] = EBADF_PAIR;
 
                                 memcpy(fds, CMSG_TYPED_DATA(found, int), sizeof(int) * 2);
 
                                 assert(mount_tree_fd < 0);
+                                assert(input_fd >= 0);
 
-                                /* Maybe we already got coredump FD in previous iteration? */
+                                /* Let's close input fd we got in the previous iteration. */
                                 safe_close(input_fd);
 
                                 input_fd = fds[0];
                                 mount_tree_fd = fds[1];
 
-                                /* We have all FDs we need let's take a shortcut here. */
+                                /* We have all FDs we need so let's take a shortcut here. */
                                 break;
-                        } else {
-                                found = cmsg_find(&mh, SOL_SOCKET, SCM_RIGHTS, CMSG_LEN(sizeof(int)));
-                                if (found)
-                                        input_fd = *CMSG_TYPED_DATA(found, int);
                         }
 
-                         /* This is the first message that carries file descriptors, maybe there will be one more that actually contains array of descriptors. */
-                        if (first) {
-                                first = false;
-                                continue;
-                        }
+                        /* This is second iteration and we didn't find array of two FDs. */
+                        found = cmsg_find(&mh, SOL_SOCKET, SCM_RIGHTS, (socklen_t) -1);
+                        cmsg_close_all(&mh);
+
+                        if (!found)
+                                /* Hence we either have no FDs which is OK and we can break. */
+                                break;
+
+                        /* Or we have some other number of FDs and somebody is playing games with us. */
+                        r = log_error_errno(SYNTHETIC_ERRNO(EBADMSG), "Received unexpected file descriptors.");
+                        goto finish;
 
```

Comment:
hmm, thinking about this more. If the protocol is extended here, I think it would make a ton of sense to extend it one step further at the same time, so that we do not have to go through the same dance a third time. Specifically, I think a 3rd fd should be passed along: the pidfd referencing the crashed process. note however, that old kernels do not support pidfds, hence it needs to be somewhat lenient on this, and support either 2 or 3 fds.

Reply:
This goes beyond fixing some style and minor issues. This requires more in-depth look and I don't have the time for it. Is this something that needs to be done befor RC1 or can it wait for Michal to be back?

Reply:
yes, it touches the protocol, I think this should be in rc1. does not have to be in this PR though, but I think it's crucial, because we go for pidfds everywhere, and it would suck if we have to do another iteration of this messy protocol augmenting again

Reply:
Ok then we can keep the issue open with rc-blocker but let's get this PR merged so that the majority of other issues are solved, including the renaming. I'll link this comment in the issue.